src/NLog/Targets/AsyncTaskTarget.cs

```diff
@@ -0,0 +1,358 @@
+                    return false;
+                }
+
+                var newTask = WriteAsyncTask(logEvent.LogEvent, _cancelTokenSource.Token);
+                if (newTask == null)
+                {
+                    InternalLogger.Debug("{0} WriteAsyncTask returned null", this.Name);
+                }
+                else
+                {
+                    lock (this.SyncRoot)
+                    {
+                        _previousTask = newTask;
+
+                        if (TaskTimeoutSeconds > 0)
+                            _taskTimeoutTimer.Change(TaskTimeoutSeconds * 1000, Timeout.Infinite);
+
+                        // NOTE - Not using _cancelTokenSource for ContinueWith, or else they will also be cancelled on timeout
+#if (SILVERLIGHT && !WINDOWS_PHONE) || NET4_0
+                        var continuation = logEvent.Continuation;
+                        _previousTask.ContinueWith(completedTask => TaskCompletion(completedTask, continuation));
+#else
+                        _previousTask.ContinueWith(_taskCompletion, logEvent.Continuation);
+#endif
+                        if (_previousTask.Status == TaskStatus.Created)
+                            _previousTask.Start(TaskScheduler);
+                    }
+                    return true;
+                }
+            }
+            catch (Exception ex)
+            {
+                try
+                {
+                    InternalLogger.Error(ex, "{0} WriteAsyncTask failed on creation", this.Name);
+                    logEvent.Continuation(ex);
+                }
+                catch
+                {
+                    // Don't wanna die
+                }
+            }
+            return false;
+        }
+
+        private void TaskCompletion(Task completedTask, object continuation)
+        {
+            try
+            {
+                if (ReferenceEquals(completedTask, _previousTask))
+                { 
+                    if (TaskTimeoutSeconds > 0)
+                    {
+                        _taskTimeoutTimer.Change(Timeout.Infinite, Timeout.Infinite);
+                    }
+                }
+                else
+                {
+                    if (!IsInitialized)
+                        return;
+                }
+
+                if (completedTask.IsCanceled)
+                {
+                    if (completedTask.Exception != null)
+                        InternalLogger.Warn(completedTask.Exception, "{0} WriteAsyncTask was cancelled", this.Name);
+                    else
+                        InternalLogger.Info("{0} WriteAsyncTask was cancelled", this.Name);
+                }
+                else if (completedTask.Exception != null)
+                {
+                    InternalLogger.Warn(completedTask.Exception, "{0} WriteAsyncTask failed on completion", this.Name);
+                }
+                ((AsyncContinuation)continuation)(completedTask.Exception);
```

Comment:
```c#
var asyncContinuation = (AsyncContinuation)continuation; 
asyncContinuation(completedTask.Exception);
```

Otherwise parenthesis mind-overflow ;)